be/dtm0_log.c

```diff
@@ -0,0 +1,631 @@
+/* -*- C -*- */
+/*
+ * Copyright (c) 2021 Seagate Technology LLC and/or its Affiliates
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+#define M0_TRACE_SUBSYSTEM M0_TRACE_SUBSYS_DTM0
+#include "be/dtm0_log.h"
+#include "be/alloc.h"
+#include "dtm0/clk_src.h"
+#include "lib/buf.h"
+#include "be/list.h"
+#include "be/op.h"
+#include "be/tx.h"
+#include "be/tx_credit.h"
+#include "be/seg.h"
+#include "lib/assert.h" /* M0_PRE */
+#include "lib/errno.h"  /* ENOENT */
+#include "lib/memory.h" /* M0_ALLOC */
+#include "lib/trace.h"
+#include "dtm0/fop.h"  /* dtm0_req_fop */
+#include "motr/magic.h"
+
+
+M0_TL_DESCR_DEFINE(lrec, "DTM0 Log", static, struct m0_dtm0_log_rec,
+                   dlr_tlink, dlr_magic, M0_BE_DTM0_LOG_REC_MAGIX,
+                   M0_BE_DTM0_LOG_MAGIX);
+M0_TL_DEFINE(lrec, static, struct m0_dtm0_log_rec);
+
+
+M0_BE_LIST_DESCR_DEFINE(lrec, "DTM0 PLog", static, struct m0_dtm0_log_rec,
+			dlr_link, dlr_magic, M0_BE_DTM0_LOG_REC_MAGIX,
+			M0_BE_DTM0_LOG_MAGIX);
+M0_BE_LIST_DEFINE(lrec, static, struct m0_dtm0_log_rec);
+
+
+static bool m0_be_dtm0_log__invariant(const struct m0_be_dtm0_log *log)
+{
+	return _0C(log != NULL) &&
+	       _0C(log->dl_cs != NULL); // &&
+	/* TODO: Add an invariant check against the volatile part */
+	       /* _0C(lrec_tlist_invariant(log->dl_tlist)); */
+}
+
+static bool m0_dtm0_log_rec__invariant(const struct m0_dtm0_log_rec *rec)
+{
+	return _0C(rec != NULL) &&
+	       _0C(m0_dtm0_tx_desc__invariant(&rec->dlr_txd)); // &&
+	       /* _0C(m0_tlink_invariant(&lrec_tl, rec)); */
+}
+
+M0_INTERNAL int m0_be_dtm0_log_init(struct m0_be_dtm0_log **out,
```

Comment:
Please add detailed comments for this function.
In which case, this function will allocate a new log.
In which case, this function call will provide a valid log by caller.

Actually, I don't like this function to be doing different things based on some flag.
We should split it into two smaller functions. One is to init a log, and another to allocate a new one.

Reply:
Done.